lib/src/editor.dart

```diff
@@ -0,0 +1,790 @@
+import 'package:yaml/src/equality.dart';
+import 'package:yaml/yaml.dart';
+
+import './source_edit.dart';
+
+/// An interface for modififying [YAML][1] documents while preserving comments and
+/// whitespaces.
+///
+/// YAML parsing is supported by `package:yaml`, and modifications are performed as
+/// string operations. Each time a modification takes place via one of the public
+/// methods, we calculate the expected final result, and parse the result YAML string,
+/// and ensure the two YAML trees match. Users may define the default settings to be
+/// applied to these string modifications. Note however that these settings only apply
+/// to portions of the YAML that are modified by this class.
+///
+/// Most modification methods require the user to pass in an [Iterable<Object>] path that
+/// holds the keys/indices to navigate to the element. Key equality is performed via
+/// `package:yaml`'s [YamlMap]'s key equality.
+///
+/// [1]: https://yaml.org/
+class YamlEditor {
+  /// List of [SourceEdit]s that have been applied to [_yaml] since the creation of this
+  /// instance, in chronological order. Intended to be compatible with `package:analysis_server`.
+  final List<SourceEdit> edits = [];
+
+  /// Current YAML string.
+  String _yaml;
+
+  /// Root node of YAML AST.
+  YamlNode _contents;
+
+  /// Style configuration settings for [YamlEditor] when modifying the YAML string.
+  final YamlStyle defaultStyle;
+
+  @override
+  String toString() => _yaml;
+
+  YamlEditor(this._yaml, {this.defaultStyle = const YamlStyle()}) {
+    _contents = loadYamlNode(_yaml);
+  }
+
+  /// Returns the [YamlNode] present at the path. The [YamlNode] that is returned represents
+  /// the current value when the function is called, and will not be updated when the YAML
+  /// is updated in the future. For example,
+  ///
+  /// ```dart
+  /// final node = doc.parseValueAt(['YAML']);
+  ///
+  /// print(node.value); /// Expected output: "YAML Ain't Markup Language"
+  ///
+  /// doc.setIn(['YAML'], 'YAML');
+  ///
+  /// final newNode = doc.parseValueAt(['YAML']);
+  ///
+  /// print(newNode.value); /// "YAML"
+  /// print(node.value); /// "YAML Ain't Markup Language"
+  /// ```
+  ///
+  /// An [ArgumentError] will be thrown if calling the `[]` operator would have resulted in an
+  /// error, but `null` (as opposed to [YamlScalar] null) will be returned if the operation would
+  /// have resulted in a `null` value on a dart collection.
+  ///
+  /// ```dart
+  /// final doc = YamlEditor('{a: {d: 4}, c: ~}');
+  /// doc.parseValueAt(['b', 'd']); // ArgumentError
+  /// doc.parseValueAt(['b']); // null
+  /// doc.parseValueAt(['c']); // YamlScalar(null)
+  ///
+  /// final doc2 = YamlEditor('[0,1]');
+  /// doc2.parseValueAt([2]); // ArgumentError
+  /// doc2.parseValueAt(["2"]); // ArgumentError
+  /// ```
+  YamlNode parseValueAt(Iterable<Object> path) {
```

Comment:
Do we really need this? Perhaps it is better to just give access to the current _contents. Then the user can parse it themselves.

Reply:
I discussed this with Jonas and we agreed that we wanted to avoid tempting the user into thinking that the `YamlNode`s will be kept updated, since in fact


keeping references to `YamlNode`s where possible, since these will be invalidated when any mutation is introduced. This was also why we used `parseValueAt`, so that the user will think that this is only valid given the YAML string at that point in time too!